### PR TITLE
[SYCL] [L0] Fix for event leakage when using immediate commandlists.

### PR DESCRIPTION
**Pull request by rdeodhar**
This change properly gathers events that have been set and need to be recycled.

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -1028,15 +1028,19 @@ _pi_queue::resetCommandList(pi_command_list_ptr_t CommandList,
...
     for (auto it = EventList.begin(); it != EventList.end(); it++) {
       std::scoped_lock<pi_shared_mutex> EventLock((*it)->Mutex);
       ze_result_t ZeResult =
           ZE_CALL_NOCHECK(zeEventQueryStatus, ((*it)->ZeEvent));
       if (ZeResult == ZE_RESULT_SUCCESS) {
-        std::move(it, it, std::back_inserter(EventListToCleanup));
-        EventList.erase(it, it);
+        EventListToCleanup.push_back(std::move((*it)));
+      } else {
+        UnsetEventsList.push_back(std::move((*it)));
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1039]: what was the reason for leak in previous code? the non-completed events remained in "EventList"

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 2]
```diff
@@ -1028,15 +1028,19 @@ _pi_queue::resetCommandList(pi_command_list_ptr_t CommandList,
...
       }
     }
+    EventList.clear();
+    EventList = std::move(UnsetEventsList);
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1043]: seems like some redundant copying, why do you need a new `UnsetEventsList` at all?

Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1043]: Trying to delete elements from EventList inside a loop which is iterating over EventList does not work. Odd things happen, like skipping elements in the original EventList. That caused the leak. It is safer to iterate over EventList, gather remaining elements in another list and then update EventList.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 1 above)
Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1039]: Yes, see comment below.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 2 above)
Comment (aelovikov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1043]: Or maybe `std::remove_if(begin, end, predicate/lambda)` ?

Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: A combination of the suggestions was used. The additional list has been eliminated and deletions from EventList are done in-place.

**Review by smaslov-intel — APPROVED** [review]

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 3]
```diff
@@ -1041,8 +1041,8 @@ _pi_queue::resetCommandList(pi_command_list_ptr_t CommandList,
       ze_result_t ZeResult =
           ZE_CALL_NOCHECK(zeEventQueryStatus, ((*it)->ZeEvent));
       if (ZeResult == ZE_RESULT_SUCCESS) {
-        std::move(it, it, std::back_inserter(EventListToCleanup));
-        EventList.erase(it, it);
+        EventListToCleanup.push_back(std::move((*it)));
+        it = EventList.erase(it, it);
```
Comment (aelovikov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1045]: What if it was the last event and `it` now points to `end` iterator? Wouldn't `it++` of it fail?

Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1045]: Even a removal at the end of the list works as expected.

Comment (aelovikov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1045]: My question wasn't about the removal but about the `it++` executed along the loop backedge after said removal is done. Is that ok as well?

Comment (rdeodhar) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: It was not OK to simply update the iterator.
Further digging revealed why in an earlier version of the code some events in the incoming list were skipped during examining for being reset. Now, the for-loop does not increment the iterator. If an event is removed the iterator is set to the element after the one being removed, and if the event is not erased then the iterator is incremented. This avoid the double ++ that was occurring before, whenever an event was removed.

**Review by aelovikov-intel — APPROVED** [review]

**Review by smaslov-intel — APPROVED** [review]

**Comment by againull** [comment]
SYCL :: AtomicRef/max_local.cpp is being tracked in https://github.com/intel/llvm/issues/7604

**Pull request merged**